provided `ffs()` and support for `popcount()` 4 and 8 byte types

psychocoderHPC opened the pull request:
- intrinsic `ffs()`
  - add test
- `popcount()` add support for any 4 and 8 byte type


@LukeTheWalker This is maybe interesting for you.

LukeTheWalker commented on tests/unit/intrinsic/ffs.cpp:
```diff
@@ -0,0 +1,174 @@
+using TestBackends
+    = std::decay_t<decltype(onHost::allBackends(onHost::enabledApis, onHost::example::enabledExecutors))>;
+
+struct PopcountKernel
```
This is probably a typo, should be something like FfsKernel

psychocoderHPC replied:
copy past issue, fixed

LukeTheWalker commented on include/alpaka/internal/intrinsic.hpp:
```diff
@@ -11,7 +11,16 @@ namespace alpaka::internal::intrinsic
         struct Op
         {
-            auto operator()(T_IntrinsicImpl const, T_Arg const& val) const;
+            int32_t operator()(T_IntrinsicImpl const, T_Arg const& val) const;
```
why did we move from auto to int32_t?

psychocoderHPC replied:
The standard is returning `int` too so I would like to be explicit. Typically we keep only `auto` if the value can be different for each implementation. This is e.g. the case for the warp function `ballot()` where the return type is 64bit in case the warp size is 64 or 32bit if the warp size is <=32

psychocoderHPC commented:
I set the PR to draft. I will remove support for types larger 8 byte. It increases the complexity without a strong requirement.
I was today a little bit overmotivated ^^

psychocoderHPC committed:
provided ffs() and support for popcount() any 4 and 8 byte type

- intrinsic `ffs()`
  - add test
- `popcount()` add support for any 4 and 8 byte type

psychocoderHPC commented:
@ikbuibui, ready to review, I reduced the PR to the necessary functionality only.

ikbuibui commented on include/alpaka/intrinsic.hpp:
```diff
@@ -9,13 +9,28 @@
+        requires(sizeof(ALPAKA_TYPEOF(arg)) == 4u || sizeof(ALPAKA_TYPEOF(arg)) == 8u)
     {
-        auto const intrinsicImpl = alpaka::trait::getIntrinsicImpl(thisApi());
+        constexpr auto const intrinsicImpl = alpaka::trait::getIntrinsicImpl(thisApi());
```
```suggestion
        constexpr auto intrinsicImpl = alpaka::trait::getIntrinsicImpl(thisApi());
```
It is not a problem but just pointing out that constexpr implies const. This also happens below.

psychocoderHPC replied:
fixed

ikbuibui reviewed (approved):
Looks good. 
As a comment I don't know if adding support for all types with matching `sizeof` by using `bit_cast` is something useful or if it can be a source of errors since the users may not expect it, but it is fine for me.

psychocoderHPC commented:
> Looks good. As a comment I don't know if adding support for all types with matching `sizeof` by using `bit_cast` is something useful or if it can be a source of errors since the users may not expect it, but it is fine for me.

I also thought about it, the STL supports only unsigned types, in alpaak mainline we already supported integral types and casted internally, therefore I saw no reason to relax it. In principle, it is a bit function and there is no reason not to allow any type.
